From e36baab578fad287ded59d76a4cf4b5366aea9b7 Mon Sep 17 00:00:00 2001
From: Dan Kephart <dan.kephart@lairdtech.com>
Date: Fri, 27 Dec 2013 13:28:17 -0500
Subject: [PATCH] Notify atheros netlink listeners that we are quitting.

Atheros events thread does not know to stop and report an error state or
restart without a 'quitting' event from the driver.  This is a
fix for that issue. This is part of the SDK events version 3 changes.

BZ4702
---
 drivers/net/wireless/ath/ath6kl/wmi.c |   15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/drivers/net/wireless/ath/ath6kl/wmi.c b/drivers/net/wireless/ath/ath6kl/wmi.c
index 128344a..8e5b379 100644
--- a/drivers/net/wireless/ath/ath6kl/wmi.c
+++ b/drivers/net/wireless/ath/ath6kl/wmi.c
@@ -4322,6 +4322,7 @@ enum {
 	ATHEROS_CMD_GET_VALUE,
 	ATHEROS_CMD_SET_PHY_MODE,
 	ATHEROS_CMD_SEND_WMI,
+	ATHEROS_CMD_QUITTING,
 	__ATHEROS_CMD_MAX,
 };
 #define ATHEROS_CMD_MAX (__ATHEROS_CMD_MAX - 1)
@@ -4637,11 +4638,25 @@ void *ath6kl_wmi_init(struct ath6kl *dev)
 
 void ath6kl_wmi_shutdown(struct wmi *wmi)
 {
+	struct sk_buff *msg;
+	void *hdr;
+
 	if (!wmi)
 		return;
  	
 	gwmi = NULL;
 
+	/* Announce our shutdown */
+	msg = genlmsg_new(NLMSG_DEFAULT_SIZE, GFP_KERNEL);
+	if (msg)
+		hdr = genlmsg_put(msg, 0, 0, &atheros_fam, 0, ATHEROS_CMD_QUITTING);
+	if (!hdr)
+		nlmsg_free(msg);
+	else {
+		genlmsg_end(msg, hdr);
+		genlmsg_multicast(msg, 0, atheros_events_mcgrp.id, GFP_KERNEL);
+	}
+
  	/* unregister the atheros family*/
  	if (genl_unregister_family(&atheros_fam) != 0) 
      	ath6kl_err("failed to unregister atheros netlink events\n");
-- 
1.7.9.5

